lib/models/project.js

```diff
@@ -409,16 +431,83 @@ class Project {
   }
 
   /**
-    Discovers all addons for this project and stores their names and
-    package.json contents in this.addonPackages as key-value pairs
-
-    @private
-    @method discoverAddons
+   * Discovers all addons for this project and stores their names and
+   * package.json contents in this.addonPackages as key-value pairs.
+   *
+   * Any packageInfos that we find that are marked as not valid are excluded.
+   *
+   * @private
+   * @method discoverAddons
    */
   discoverAddons() {
-    let addonsList = this.addonDiscovery.discoverProjectAddons(this);
+    let pkgInfo = this.packageInfoCache.getEntry(this.root);
```

Comment:
`Project.prototype.discoverAddons` and `Addon.prototype.discoverAddons` are both complex functions and seem similar. Could we unify them as a single (or a handful) standalone functions? The duplication between comes at a high risk/cost currently.

This should mitigate the last technical concern @rwjblue has.

Reply:
Guys, I think you're basically asking me to unify two things that are not the same, and in a way that doesn't really make sense. Because they aren't the same, and because they take advantage of data that is specific to the particular objects, the idea of encapsulation dictates that they really should belong where they are. Here's a pseudo-code version of each:

```javascript
Project.discoverAddons() {
    get package entry
    create empty addonPackageList
    if this is an emberCLIAddon, add it to the package list
    if the project has a CLI, add those in-repo addons to the list
    if the project has internal addons, add those to the list
    if the package has 'dependencies' packages, add any that are addons to the list
    if the package has 'devDependencies' packages, add any that are addons to the list
    if the package itself has in-repo addons, add those to the list
    create an object with an AddonInfo for each valid addon package
    write out error messages for any invalid addon packages.
}
```
```javascript
Addon.discoverAddons() {
    get package entry
    create empty addon package list
    if we actually have package information (the addon is in the cache) {
        add any 'dependencies' packages that are NOT 'ember-cli' to the package list
     }
    add in-repo addons to the lsit

    create the same object of AddonInfo instances, except for each addon package,
        make sure that the calling addon allows it.

    write to the console for any addons that are considered invalid (
}
```

Given that project has many more places to check, and Addon has various checks for things like shouldIncludeChildAddon, the actual number of duplicate lines in the two routines is maybe 10 or 12. Having a combined routine externally that has to reach into the Project object or Addon object, or take several parameters to distinguish between them, is just bad code, in my opinion. It violates encapsulation and requires extra processing where none is needed, slowing things down (however slightly). 

I really feel strongly that these two should stay where they are for those reasons. The AddonDiscovery had the same sorts of problems, which is part of why it should have been refactored in the first place.

Reply:
Changes are done.